### ISSUE TITLE
`const_evaluatable_unchecked` doesn't correctly deal with non default substs

### ISSUE THREAD

**Issue by lcnr** [rust-lang/rust#94293]
While checking whether the [fixed crate](https://crates.io/crates/fixed) can be ported from typenum to nightly with `generic_const_exprs`, I hit this warning. I am not sure whether it's a false positive or whether the behavior I want is meant to be disallowed, but I think it should be allowed when enabling `generic_const_exprs`.

I've reduced the code though I left some stuff to show why it would be useful.
```rust
#![feature(generic_const_exprs)]

pub struct If<const CONDITION: bool>;
pub trait True {}
impl True for If<true> {}

pub struct FixedI8<const FRAC: u32> {
    pub bits: i8,
}

impl<const FRAC_LHS: u32, const FRAC_RHS: u32> PartialEq<FixedI8<FRAC_RHS>> for FixedI8<FRAC_LHS>
where
    If<{ FRAC_RHS <= 8 }>: True,
{
    fn eq(&self, _rhs: &FixedI8<FRAC_RHS>) -> bool {
        unimplemented!()
    }
}

impl<const FRAC: u32> PartialEq<i8> for FixedI8<FRAC> {
    fn eq(&self, rhs: &i8) -> bool {
        let rhs_as_fixed = FixedI8::<0> { bits: *rhs };
        PartialEq::eq(self, &rhs_as_fixed)
    }
}
```

_Originally posted by @tspiteri in https://github.com/rust-lang/rust/issues/76200#issuecomment-1048580007_

**Comment by lcnr** [comment]
`const_evaluatable_unchecked` is emitted if we successfully evaluate an anonymous constant with generic const arguments.

Anonymous constants currently inherit all of their parents generics, so for `If<{ FRAC_RHS <= 8 }>: True,` the constant `{ FRAC_RHS <= 8 }` has both `FRAC_RHS` and `FRAC_LHS` in scope, even though it only uses `FRAC_RHS`.

This would mean that pretty much all anonymous constants would cause that lint, so we add a flag to their MIR to detect whether they actually use any generic parameter: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/struct.Body.html#structfield.is_polymorphic.

`{ FRAC_RHS <= 8 }` uses `FRAC_RHS` so that flag is true. The problem is now that `FRAC_RHS` is concrete, so evaluating the constant is successful, but the constant refers to `FRAC_LHS` which is still generic.

We should just disable the `const_evaluatable_unchecked` lint if `feature(generic_const_exprs)` is enabled :shrug: 

cc @BoxyUwU

**Comment by peter-kehl** [comment]
Hi. I've been referred to here from #76200. I didn't fully understand #76200, and the above is even more advanced.

I have use case for an associative type with a bound that depends on its owner (trait)'s const generic parameter. Any alternatives/workarounds, please?

Short example failing on Nightly (thanks to ilyvion#9946 for narrowing it down): https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=c133c09b2fa155bb3f4b9ee45e6c011e:
```rust
#![feature(generic_const_exprs)]

trait Foo<const N: usize>
where
    [(); check_empty_array_size(N)]: Sized,
{
    type Narr: Foo<0>;
}

#[allow(unused_variables)]
pub const fn check_empty_array_size(array_size: usize) -> usize {
    0
}
```

My use case: https://github.com/ranging-rs/slicing-rs/blob/main/src/slices/mod.rs#L30=. I know it's a semantic/voluntary constraint only, and it's quite recursive, but I believe it deserves its place. (Alternatives I tried are at https://github.com/ranging-rs/slicing-rs/issues/1.)

**Comment by lcnr** [comment]
this has been fixed, needs test to close
